### ISSUE TITLE
MiniBatch size and convergence

### ISSUE THREAD

**Issue by JohnCraigPublic** [Microsoft/CNTK#1621]
I have a network that converges fine if minibatch size is set to 64.  If I set it to 128 then errors never reduce -- it does not converge.  Does this make sense?   If minibatch size is increased do I need to lower learning rate to compensate?

**Comment by cha-zhang** [comment]
Yes, you do. CNTK uses sum of gradient rather than average gradient for gradient update. Another possibility is to use learning rate per sample instead of per minibatch (e.g., learning_rate_schedule(0.0005, UnitType.sample)), which is actually the recommended way in CNTK.

**Comment by JohnCraigPublic** [comment]
Aha! Thanks for the answer!

**Comment by arijit17** [comment]
Hi @cha-zhang: 

Just a question.
Is it possible as a user to choose between sum gradient or average gradient?

**Comment by tangyuq** [comment]
Yes, if the minibatch_size is set to 1 in the learner, it recovers the sum gradients behaviors. 

For examples, in the following learners, the first one use average gradients and the second one use sum gradients:
sgd_learner_average = C.sgd(z.parameters, lr = 0.5, minibatch_size = C.learners.IGNORE)
sgd_learner_sum = C.sgd(z.parameters, lr = 0.5, minibatch_size = 1)

Actually, other than setting to IGNORE, the learner is on sum gradient with learning rate being automatically adjusted for sum gradients. The learning rate adjustment is to simulate the behavior of the specified learning rate on the average gradient of specified minibatch size (reference minibatch size) while it is actually on sum gradients. In this way, CNTK can have the advantage of using sum gradients for distributed training.   See [use learners manual](https://github.com/Microsoft/CNTK/blob/master/Manual/Manual_How_to_use_learners.ipynb) for more information.
